code/modules/admin/bccm/bccm.dm

```diff
@@ -0,0 +1,387 @@
+		row["timestamp"] = _Whitelist_DB_Select_Query.item[3]
+
+		result["displayData"] += list(row)
+
+	qdel(_Whitelist_DB_Select_Query)
+
+	return result
+
+/datum/controller/subsystem/bccm/proc/AddASNban(address, client/Admin)
+	if(!CheckDBCon())
+		return
+
+	if(!check_rights(R_SERVER, TRUE, Admin))
+		return
+
+	var/ip = remove_all_spaces(new_sql_sanitize_text(address))
+
+	if(length(ip) > 16)
+		return
+
+	var/list/response = GetAPIresponse(ip)
+
+	var/ip_as = response["as"]
+
+	var/datum/db_query/_ASban_Insert_Query = SSdbcore.NewQuery("INSERT INTO bccm_asn_ban (`ip`, `asn`, `a_ckey`, `timestamp`) VALUES ('[ip]', '[ip_as]', '[Admin.ckey]', Now())")
+	_ASban_Insert_Query.Execute()
+	qdel(_ASban_Insert_Query)
+
+	tgui_panel_asn_data = GetAsnBanlistDatabase()
+	log_and_message_admins("has added '[ip_as]' to the BCCM ASN banlist.", Admin)
+
+	return TRUE
+
+/datum/controller/subsystem/bccm/proc/RemoveASNban(ip_as, client/Admin)
+	if(!CheckDBCon())
+		return
+
+	if(!check_rights(R_SERVER, TRUE, Admin))
+		return
+
+	var/datum/db_query/_ASban_Delete_Query = SSdbcore.NewQuery("DELETE FROM bccm_asn_ban WHERE `asn` = '[ip_as]'")
+	_ASban_Delete_Query.Execute()
+	qdel(_ASban_Delete_Query)
+
+	tgui_panel_asn_data = GetAsnBanlistDatabase()
+	log_and_message_admins("has removed '[ip_as]' from the BCCM ASN banlist.", Admin)
+
+	return TRUE
+
+
+/datum/controller/subsystem/bccm/proc/GetAsnBanlistDatabase()
+	var/datum/db_query/_ASN_Banlist_Select_Query = SSdbcore.NewQuery("SELECT `asn`, `timestamp`, `a_ckey` from bccm_asn_ban")
+	_ASN_Banlist_Select_Query.Execute()
+
+	var/list/result = list()
+
+	while(_ASN_Banlist_Select_Query.NextRow())
+		var/list/row = list()
+		row["asn"] = _ASN_Banlist_Select_Query.item[1]
+		row["timestamp"] = _ASN_Banlist_Select_Query.item[2]
+		row["a_ckey"] = _ASN_Banlist_Select_Query.item[3]
+
+		result["displayData"] += list(row)
+
+	qdel(_ASN_Banlist_Select_Query)
+
+	return result
+
+
+/datum/controller/subsystem/bccm/proc/HandleClientAccessCheck(client/C, postponed = 0)
+	if(!SSbccm.CheckForAccess(C) && !(C.ckey in GLOB.admin_datums))
+		if(!postponed)
+			C.log_client_to_db_connection_log()
+		log_and_message_admins(SPAN_NOTICE("BCCM: Failed Login: [C.key]/[C.ckey]([C.address])([C.computer_id]) failed to pass BCCM check."))
+		qdel(C)
+		return
+
+/datum/controller/subsystem/bccm/proc/HandleASNbanCheck(client/C, postponed = 0)
+	if(!SSbccm.CheckASNban(C) && !(C.ckey in GLOB.admin_datums))
+		if(!postponed)
+			C.log_client_to_db_connection_log()
+		log_and_message_admins(SPAN_NOTICE("BCCM: Failed Login: [C.key]/[C.ckey]([C.address])([C.computer_id]) failed to pass ASN ban check."))
+		qdel(C)
+		return
+
+/client/proc/BCCM_toggle()
+	set category = "Server"
+	set name = "Toggle BCCM"
+
+	if(!SSdbcore.Connect())
+		to_chat(usr, SPAN_NOTICE("The Database is not connected!"))
+		return
+
+	var/bccm_status = SSbccm.Toggle()
+	log_and_message_admins("has [bccm_status ? "enabled" : "disabled"] the BCCM system!")
```

Comment:
You might want to double check and make sure the client is an admin and can do this here